### PR TITLE
0-dim tensor support

### PR DESCRIPTION
**Pull request by Flowingsun007**

### CODE REVIEW

File: oneflow/api/python/framework/shape.cpp [hunk 1]
```diff
@@ -25,6 +25,7 @@ namespace oneflow {
 namespace {
 struct ShapeExportUtil final {
   static Maybe<Shape> MakeShape(const py::tuple& py_shape) {
+    if (py_shape.empty()) { return std::make_shared<Shape>(); }
```
Comment (lixinqi) [inline oneflow/api/python/framework/shape.cpp:28]: 这里显然错了吧。
应该是`std::make_shared<Shape>(DimVector{});`
否者那个elem_cnt不可访问。

File: oneflow/api/python/framework/shape.cpp [hunk 2]
```diff
@@ -66,9 +67,13 @@ struct ShapeExportUtil final {
     std::stringstream ss;
     int32_t idx = 0;
     ss << "flow.Size([";
-    for (int64_t dim : shape.dim_vec()) {
-      ss << dim;
-      if (++idx != shape.dim_vec().size()) { ss << ", "; }
+    if (!shape.is_uninitialized()) {
```
Comment (lixinqi) [inline oneflow/api/python/framework/shape.cpp:70]: 去掉这个if。因为上边的代码就不应该产生一个未初始化的形状。

File: oneflow/core/common/shape.cpp [hunk 3]
```diff
@@ -93,9 +91,11 @@ std::string Shape::ToString() const {
   std::stringstream ss;
   int32_t idx = 0;
   ss << "(";
-  for (int64_t dim : dim_vec_) {
-    ss << dim;
-    if (++idx != dim_vec_.size() || dim_vec_.size() == 1) { ss << ","; }
+  if (!is_uninitialized()) {
```
Comment (lixinqi) [inline oneflow/core/common/shape.cpp:94]: 去掉

File: oneflow/api/python/framework/shape.cpp (hunk 2 above)
Comment (Flowingsun007) [reply oneflow/api/python/framework/shape.cpp:70]: 好的

File: oneflow/python/test/modules/test_0_dim_tensor.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+"""
```
Comment (lixinqi) [inline oneflow/python/test/modules/test_0_dim_tensor.py:1]: 这份文件路径错了，应该是python/oneflow/test/modules/test_0_dim_tensor.py

Comment (Flowingsun007) [reply oneflow/python/test/modules/test_0_dim_tensor.py]: 已修改

File: oneflow/core/common/shape.cpp (hunk 3 above)
Comment (Flowingsun007) [reply oneflow/core/common/shape.cpp]: done

File: oneflow/api/python/framework/shape.cpp (hunk 1 above)
Comment (Flowingsun007) [reply oneflow/api/python/framework/shape.cpp]: done

File: oneflow/api/python/framework/shape.cpp [hunk 5]
```diff
@@ -66,9 +67,11 @@ struct ShapeExportUtil final {
     std::stringstream ss;
     int32_t idx = 0;
     ss << "flow.Size([";
-    for (int64_t dim : shape.dim_vec()) {
-      ss << dim;
-      if (++idx != shape.dim_vec().size()) { ss << ", "; }
+    if (shape.dim_vec().size() > 0) {
```
Comment (lixinqi) [inline oneflow/api/python/framework/shape.cpp:70]: 去掉这个if。没有必要

File: oneflow/core/common/shape.h [hunk 6]
```diff
@@ -54,11 +55,15 @@ class Shape final {
...
   // Getters and Setters
+  bool is_uninitialized() const { return !elem_cnt_.has_value(); }
   const DimVector& dim_vec() const { return dim_vec_; }
-  int64_t elem_cnt() const { return elem_cnt_; }
+  int64_t elem_cnt() const { return CHECK_JUST(elem_cnt_.value()); }
   int64_t At(int64_t index) const { return dim_vec_.at(index); }
   void Set(int64_t index, int64_t val);
-  int64_t NumAxes() const { return dim_vec_.size(); }
+  int64_t NumAxes() const {
+    if (is_uninitialized()) { return 0; }
```
Comment (lixinqi) [inline oneflow/core/common/shape.h:64]: CHECK(!is_uninitialized());

File: python/oneflow/nn/modules/constant.py [hunk 7]
```diff
@@ -34,15 +34,18 @@ def __init__(
...
         self.requires_grad = requires_grad
         size = _single(size)
         if dtype is None:
             dtype = flow.float32
         if device is None:
             self.device = flow.device("cpu")
+        elif isinstance(device, str):
+            self.device = flow.device(device)
+        else:
+            self.device = device
```
Comment (lixinqi) [inline python/oneflow/nn/modules/constant.py:48]: assert isinstance(device, flow.device)

File: python/oneflow/framework/tensor.py [hunk 8]
```diff
@@ -768,7 +776,11 @@ def _construct_with_data(
         if _input_args_is_tuple_or_list(*args):
             numpy_data = np.array(args[0])
         elif _input_args_is_numpy(*args):
-            numpy_data = np.ascontiguousarray(args[0])
+            if args[0].size == 1 and args[0].ndim == 0:
+                numpy_data = args[0]
+            else:
+                numpy_data = np.ascontiguousarray(args[0])
```
Comment (lixinqi) [inline python/oneflow/framework/tensor.py:782]: 没看明白这里为什么要特判？难道0-D张量就不能ascontiguousarray了吗？

File: python/oneflow/framework/tensor.py [hunk 9]
```diff
@@ -58,8 +58,13 @@ def _local_tensor_numpy(eager_local_tensor):
         return [t.numpy() for t in tensors]
     method_name = eager_local_tensor._get_copy_mirrored_tensor_to_numpy_func_name()
     copy_to_numpy = getattr(eager_local_tensor, method_name)
+
+    n_shape = (
+        () if len(eager_local_tensor.shape) == 0 else tuple(eager_local_tensor.shape)
+    )
```
Comment (lixinqi) [inline python/oneflow/framework/tensor.py:64]: 为什么又有特判？

File: oneflow/core/common/shape.h [hunk 10]
```diff
@@ -54,11 +55,15 @@ class Shape final {
   void SerializeWithTextFormat(StreamT& out_stream) const;
 
   // Getters and Setters
+  bool is_uninitialized() const { return !elem_cnt_.has_value(); }
```
Comment (lixinqi) [inline oneflow/core/common/shape.h:58]: 是不是不要写成否定词，而是`is_initialized()`

File: oneflow/core/common/shape.h (hunk 6 above)
Comment (Flowingsun007) [reply oneflow/core/common/shape.h]: 已修改

File: oneflow/api/python/framework/shape.cpp (hunk 5 above)
Comment (Flowingsun007) [reply oneflow/api/python/framework/shape.cpp]: 已修改

File: python/oneflow/framework/tensor.py [hunk 11]
```diff
@@ -73,10 +78,10 @@ def _copy_from_numpy_to_eager_local_tensor(eager_local_tensor, np_arr):
...
         eager_local_tensor.dtype
     )
-    if np_arr.shape == ():
-        assert tuple(eager_local_tensor.shape) == (1,)
-    else:
-        assert np_arr.shape == tuple(eager_local_tensor.shape)
+    # if np_arr.shape == ():
+    #     assert tuple(eager_local_tensor.shape) == (1,)
+    # else:
+    #     assert np_arr.shape == tuple(eager_local_tensor.shape)
```
[the remark this answers is not in the document]
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py:84]: 好的

File: python/oneflow/framework/tensor.py (hunk 8 above)
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py:782]: 应该也是可以的，不过感觉没必要

File: oneflow/core/common/shape.h (hunk 10 above)
Comment (Flowingsun007) [reply oneflow/core/common/shape.h:58]: 嗯，感觉is_initialized语意更好点

File: python/oneflow/framework/tensor.py (hunk 9 above)
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py:64]: 这块之前出错了，加了判断才正常（现在应该不需要这样特判了，我试试

File: python/oneflow/framework/tensor.py (hunk 8 above)
Comment (lixinqi) [reply python/oneflow/framework/tensor.py:782]: 到底什么是“没有必要”？你这里难道在担心机器对边角情况干了多余的事情吗？
最大的“没有必要”是一个又一个毫无意义的特判。review起来不舒服，维护起来不舒服。

File: python/oneflow/nn/modules/constant.py (hunk 7 above)
Comment (Flowingsun007) [reply python/oneflow/nn/modules/constant.py:48]: device也可以传str类型的

File: python/oneflow/framework/tensor.py (hunk 8 above)
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py:782]: 好的，我去掉if判断吧

File: python/oneflow/nn/modules/constant.py (hunk 7 above)
Comment (lixinqi) [reply python/oneflow/nn/modules/constant.py:48]: str的情况上一个if分支已经处理了

Comment (Flowingsun007) [reply python/oneflow/nn/modules/constant.py:48]: 好的

File: python/oneflow/framework/tensor.py (hunk 8 above)
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py:778]: 这个不能去掉，0-d的scalar如果用ascontiguousarray处理后，会变成[x]这样的，维度为1

File: oneflow/core/common/shape.h (hunk 10 above)
Comment (Flowingsun007) [reply oneflow/core/common/shape.h]: done

File: python/oneflow/framework/tensor.py (hunk 9 above)
Comment (Flowingsun007) [reply python/oneflow/framework/tensor.py]: 已修改

File: oneflow/core/common/shape.cpp [hunk 12]
```diff
@@ -190,7 +187,7 @@ Maybe<Shape> Shape::Slice(int64_t start_dim, int64_t end_dim) const {
   if (start_dim > ndims) { start_dim = ndims; }
   if (end_dim > ndims) { end_dim = ndims; }
   DimVector dim_vec;
-  for (int64_t i = start_dim; i < end_dim && i < ndims; ++i) { dim_vec.push_back(this->At(i)); }
```
Comment (lixinqi) [inline oneflow/core/common/shape.cpp:193]: 以后养成用++i的习惯。i++会有一些非常隐晦的问题。

**Review by lixinqi — APPROVED** [review]

File: oneflow/core/common/shape.cpp (hunk 12 above)
Comment (Flowingsun007) [reply oneflow/core/common/shape.cpp]: 好的

File: oneflow/api/python/framework/shape.cpp [hunk 13]
```diff
@@ -25,6 +25,7 @@ namespace oneflow {
 namespace {
 struct ShapeExportUtil final {
   static Maybe<Shape> MakeShape(const py::tuple& py_shape) {
+    if (py_shape.empty()) { return std::make_shared<Shape>(DimVector{}); }
```
Comment (lixinqi) [inline oneflow/api/python/framework/shape.cpp:28]: 去掉。因为下面的代码处理了这种情形

Comment (Flowingsun007) [reply oneflow/api/python/framework/shape.cpp]: 好的，已删
